namesys/routing.go

```diff
@@ -76,59 +88,109 @@ func (r *IpnsResolver) resolveOnce(ctx context.Context, name string, options *op
-		return "", 0, err
+		log.Debugf("RoutingResolver: dht get for name %s failed: %s", name, err)
+		out <- onceResult{err: err}
+		close(out)
+		cancel()
+		return out
 	}
 
-	var p path.Path
-	// check for old style record:
-	if valh, err := mh.Cast(entry.GetValue()); err == nil {
-		// Its an old style multihash record
-		log.Debugf("encountered CIDv0 ipns entry: %s", valh)
-		p = path.FromCid(cid.NewCidV0(valh))
-	} else {
-		// Not a multihash, probably a new record
-		p, err = path.ParsePath(string(entry.GetValue()))
-		if err != nil {
-			return "", 0, err
+	go func() {
+		defer cancel()
+		defer close(out)
+		for {
+			select {
+			case val, ok := <-vals:
+				if !ok {
+					return
+				}
+
+				entry := new(pb.IpnsEntry)
+				err = proto.Unmarshal(val, entry)
+				if err != nil {
+					log.Debugf("RoutingResolver: could not unmarshal value for name %s: %s", name, err)
+					select {
+					case out <- onceResult{err: err}:
+					case <-ctx.Done():
+					}
+					return
+				}
+
+				var p path.Path
+				// check for old style record:
+				if valh, err := mh.Cast(entry.GetValue()); err == nil {
+					// Its an old style multihash record
+					log.Debugf("encountered CIDv0 ipns entry: %s", valh)
+					p = path.FromCid(cid.NewCidV0(valh))
+				} else {
+					// Not a multihash, probably a new style record
+					p, err = path.ParsePath(string(entry.GetValue()))
+					if err != nil {
+						select {
+						case out <- onceResult{err: err}:
+						case <-ctx.Done():
+						}
+						return
+					}
+				}
+
+				ttl := DefaultResolverCacheTTL
+				if entry.Ttl != nil {
+					ttl = time.Duration(*entry.Ttl)
+				}
+				switch eol, err := ipns.GetEOL(entry); err {
+				case ipns.ErrUnrecognizedValidity:
+					// No EOL.
+				case nil:
+					ttEol := eol.Sub(time.Now())
+					if ttEol < 0 {
+						// It *was* valid when we first resolved it.
+						ttl = 0
+					} else if ttEol < ttl {
+						ttl = ttEol
+					}
+				default:
+					log.Errorf("encountered error when parsing EOL: %s", err)
+					select {
+					case out <- onceResult{err: err}:
+					case <-ctx.Done():
+					}
+					return
+				}
+
+				select {
+				case out <- onceResult{value: p, ttl: ttl}:
+				case <-ctx.Done():
+				}
+			case <-ctx.Done():
+				return
+			}
 		}
-	}
+	}()
 
-	ttl := DefaultResolverCacheTTL
-	if entry.Ttl != nil {
-		ttl = time.Duration(*entry.Ttl)
-	}
-	switch eol, err := ipns.GetEOL(entry); err {
-	case ipns.ErrUnrecognizedValidity:
-		// No EOL.
-	case nil:
-		ttEol := eol.Sub(time.Now())
-		if ttEol < 0 {
-			// It *was* valid when we first resolved it.
-			ttl = 0
-		} else if ttEol < ttl {
-			ttl = ttEol
-		}
-	default:
-		log.Errorf("encountered error when parsing EOL: %s", err)
-		return "", 0, err
-	}
+	return out
+}
 
-	return p, ttl, nil
+func (r *IpnsResolver) searchValue(ctx context.Context, key string, opts ...ropts.Option) (<-chan []byte, error) {
+	if ir, ok := r.routing.(*dht.IpfsDHT); ok {
+		return ir.SearchValue(ctx, key, opts...)
```

Comment:
Mind:

1. Fixing the interfaces.
2. Taking over https://github.com/libp2p/go-libp2p-pubsub-router/pull/3 (or start over), making it follow the new API.

Instead?

Reply:
> Taking over libp2p/go-libp2p-pubsub-router#3 (or start over), making it follow the new API.

I wanted to do that, but in a separate PR

Reply:
Yeah, as-is, this'll fail if ipns-over-pubsub is enabled. We could possibly work around that but I'd rather not.